libsignal-service/src/cipher.rs

```diff
@@ -422,3 +443,106 @@ pub async fn get_preferred_protocol_address<S: SessionStore>(
 
     Ok(ProtocolAddress::new(address.identifier(), device_id))
 }
+
+/// Decrypt a Sealed Sender message `ciphertext` in either the v1 or v2 format, validate its sender
+/// certificate, and then decrypt the inner message payload.
+///
+/// This method calls [`sealed_sender_decrypt_to_usmc`] to extract the sender information, including
+/// the embedded [`SenderCertificate`]. The sender certificate (signed by the [`ServerCertificate`])
+/// is then validated against the `trust_root` baked into the client to ensure that the sender's
+/// identity was not forged.
+#[allow(clippy::too_many_arguments)]
+async fn sealed_sender_decrypt(
```

Comment:
I don't immediately see how this works. Should the application call this method when the `ContentBody` matches some specific variant? This function is, as far as I can see, never used (but clippy doesn't call you out on it, so I'm potentially blind).

Since you tested in Presage, I assume this works, for both receiving and sending.

Reply:
This is (unfortunately) a copy-pasta of the current impl in `libsignal` and is used here https://github.com/whisperfish/libsignal-service-rs/blob/34e24634e2bb66b48f703f619acaed870a3cd475/libsignal-service/src/cipher.rs#L219 the same way we should use the function in `libsignal`, which somehow doesn't have a branch for `CiphertextMessageType::SenderKey` https://github.com/signalapp/libsignal/blob/main/rust/protocol/src/sealed_sender.rs#L1647

Reply:
I don't quite understand why this is another function. I checked the sources of Signal Desktop. And there this code is part of `open_envelope`.

Reply:
Basically, what they do there: 

1. Unseal the envelope with `sealedSenderDecryptToUsmc`
2. Check the `msgtype` (here is should be part of case for unidentified sender)
3. Depending on the type of `msgtype` call `decrypt` or `groupDecrypt`

Reply:
Ah, now I got it. Basically, you are doing it one level deeper compared to Signal Desktop, which actually makes sense.